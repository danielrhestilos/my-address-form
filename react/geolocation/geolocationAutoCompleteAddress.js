import flow from 'lodash/flow'

import getCountryISO2 from '../countryISOMap'
import {
  addNewField,
  addFocusToNextInvalidField,
  createNewAddress,
} from '../transforms/address'

export default function geolocationAutoCompleteAddress(
  baseAddress,
  googleAddress,
  rules
) {
  const geolocationRules = rules.geolocation
  const fallbackCountry = rules.country

  let address = flow([
    setAddressFields,
    runGeolocationFieldHandlers,
    setGeoCoordinates,
    setCountry,
    setAddressQuery,
    (updatedAddress) =>
      addNewField(
        updatedAddress,
        'geolocationAutoCompleted',
        (fieldValue) => fieldValue?.value !== null
      ),
    (updatedAddress) => ({
      ...updatedAddress,
      addressId: baseAddress.addressId,
      addressType: baseAddress.addressType,
      receiverName: baseAddress.receiverName,
    }),
    (updatedAddress) => addFocusToNextInvalidField(updatedAddress, rules),
  ])()

  // Set notApplicable to number if property there is in geolocationRules
  address = setNumberNotApplicable(address)

  return address

  // The functions below use googleAddress and geolocationRules
  // from the closure created.

  function setAddressFields() {
    return Object.entries(geolocationRules).reduce(
      (updatedAddress, [fieldName, fieldValue]) => {
        const component = getAddressComponent(
          googleAddress.address_components,
          fieldValue.types
        )

        if (component) {
          updatedAddress = setAddressFieldValue(
            updatedAddress,
            fieldName,
            geolocationRules,
            component
          )
        }

        return updatedAddress
      },
      createNewAddress()
    )
  }

  function setGeoCoordinates(updatedAddress) {
    const { location } = googleAddress.geometry

    updatedAddress.geoCoordinates = {
      visited: true,
      value: [
        typeof location.lng === 'function' ? location.lng() : location.lng,
        typeof location.lat === 'function' ? location.lat() : location.lat,
      ],
    }

    return updatedAddress
  }

  // Run custom function handlers to fill some fields
  function runGeolocationFieldHandlers(updatedAddress) {
    const ruleEntries = Object.entries(geolocationRules)

    ruleEntries.forEach(([, rule]) => {
      if (rule.handler) {
        updatedAddress = rule.handler(updatedAddress, googleAddress)
      }
    })

    ruleEntries.forEach(([, rule]) => {
      if (rule.handler) {
        updatedAddress = rule.handler(updatedAddress, googleAddress)
      }
    })

    return updatedAddress
  }

  function setCountry(updatedAddress) {
    const country = getCountry(googleAddress)

    updatedAddress.country = { value: country || fallbackCountry }

    return updatedAddress
  }

  function setAddressQuery(updatedAddress) {
    updatedAddress.addressQuery = { value: googleAddress.formatted_address }

    return updatedAddress
  }

  function setNumberNotApplicable(updatedAddress) {
    const geolocationNumberCondition =
      geolocationRules &&
      geolocationRules.number &&
      geolocationRules.number.notApplicable

    if (geolocationNumberCondition) {
      return {
        ...updatedAddress,
        number: {
          ...updatedAddress.number,
          notApplicable: true,
        },
      }
    }

    return updatedAddress
  }
}

function getAddressComponent(addressComponents, types) {
  for (const type of types ?? []) {
    const addressComponent = addressComponents.find((component) =>
      component.types.includes(type)
    )

    if (addressComponent) {
      return addressComponent
    }
  }

  return undefined
}

function setAddressFieldValue(
  address,
  fieldName,
  geolocationRules,
  addressComponent
) {
  const geolocationField = geolocationRules[fieldName]

  address[fieldName] = { value: addressComponent[geolocationField.valueIn] }

  return address
}

function getCountry(googleAddress) {
  const countryComponent = googleAddress.address_components.find(
    (component) => component.types.indexOf('country') !== -1
  )

  return countryComponent ? getCountryISO2(countryComponent.short_name) : null
}
