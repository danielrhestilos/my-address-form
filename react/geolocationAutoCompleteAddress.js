import geolocationAutoCompleteAddress from './geolocation/geolocationAutoCompleteAddress'

export default { geolocationAutoCompleteAddress }
