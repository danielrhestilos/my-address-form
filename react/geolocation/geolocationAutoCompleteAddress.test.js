import geolocationAutoCompleteAddress from './geolocationAutoCompleteAddress'
import postalCodeGoogleAddress from './__mocks__/postalCodeGoogleAddress'
import postalCodeGoogleAddressNoNumber from './__mocks__/postalCodeGoogleAddressNoNumber'
import oneLevelGoogleAddress from './__mocks__/oneLevelGoogleAddress'
import invalidFieldGoogleAddress from './__mocks__/invalidFieldGoogleAddress'
import usePostalCode from '../country/__mocks__/usePostalCode'
import useOneLevel from '../country/__mocks__/useOneLevel'
import newAddress from '../__mocks__/newAddress'

describe('Geolocation Auto Complete Address', () => {
  it('should transform a Google address to a Checkout address', () => {
    const address = geolocationAutoCompleteAddress(
      newAddress,
      postalCodeGoogleAddress,
      usePostalCode
    )

    expect(address).toMatchSnapshot()
  })

  it('should call handlers to fill postalCode', () => {
    const address = geolocationAutoCompleteAddress(
      newAddress,
      oneLevelGoogleAddress,
      useOneLevel
    )

    expect(address.postalCode.value).toBe('0000')
  })

  it('should get an address with a focus on the first invalid field', () => {
    const address = geolocationAutoCompleteAddress(
      newAddress,
      invalidFieldGoogleAddress,
      usePostalCode
    )

    expect(address.postalCode.valid).toBe(false)
    expect(address.postalCode.focus).toBe(true)
  })

  it('should keep addressId, addressType and receiverName', () => {
    const addressId = '1'
    const addressType = 'residential'
    const receiverName = 'Linus'

    const address = geolocationAutoCompleteAddress(
      {
        ...newAddress,
        addressId: { value: addressId },
        addressType: { value: addressType },
        receiverName: { value: receiverName },
      },
      postalCodeGoogleAddress,
      usePostalCode
    )

    expect(address.addressId.value).toBe(addressId)
    expect(address.addressType.value).toBe(addressType)
    expect(address.receiverName.value).toBe(receiverName)
  })

  it('should empty fields that are not auto completed', () => {
    const complement = 'apt 505'

    const address = geolocationAutoCompleteAddress(
      {
        ...newAddress,
        complement: { value: complement },
      },
      postalCodeGoogleAddress,
      usePostalCode
    )

    expect(address.complement.value).toBeNull()
  })

  it('should keep number as notApplicable', () => {
    const address = geolocationAutoCompleteAddress(
      { ...newAddress },
      postalCodeGoogleAddressNoNumber,
      usePostalCode
    )

    expect(address.number).toBeTruthy()
    expect(address.number.notApplicable).toBeTruthy()
  })
})
