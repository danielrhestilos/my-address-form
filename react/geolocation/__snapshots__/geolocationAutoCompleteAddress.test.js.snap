// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Geolocation Auto Complete Address should transform a Google address to a Checkout address 1`] = `
Object {
  "addressId": Object {
    "value": "1",
  },
  "addressQuery": Object {
    "geolocationAutoCompleted": true,
    "value": "R. Barão de Itambí, 55 - Botafogo, Rio de Janeiro - RJ, Brasil",
  },
  "addressType": Object {
    "value": "residential",
  },
  "city": Object {
    "geolocationAutoCompleted": true,
    "value": "Rio de Janeiro",
  },
  "complement": Object {
    "geolocationAutoCompleted": false,
    "value": null,
  },
  "country": Object {
    "geolocationAutoCompleted": true,
    "value": "BRA",
  },
  "geoCoordinates": Object {
    "geolocationAutoCompleted": true,
    "value": Array [
      -43.18075109999995,
      -22.941129,
    ],
    "visited": true,
  },
  "neighborhood": Object {
    "geolocationAutoCompleted": true,
    "value": "Botafogo",
  },
  "number": Object {
    "geolocationAutoCompleted": true,
    "notApplicable": true,
    "value": "55",
  },
  "postalCode": Object {
    "geolocationAutoCompleted": true,
    "value": "22231000",
  },
  "receiverName": Object {
    "focus": true,
    "geolocationAutoCompleted": undefined,
    "reason": "ERROR_EMPTY_FIELD",
    "valid": false,
    "value": null,
    "valueOptions": undefined,
  },
  "reference": Object {
    "geolocationAutoCompleted": false,
    "value": null,
  },
  "state": Object {
    "geolocationAutoCompleted": true,
    "value": "RJ",
  },
  "street": Object {
    "geolocationAutoCompleted": true,
    "value": "Rua Barão de Itambí",
  },
}
`;
